File: BackEnd/Domains/ShoppingCart/CreateCart/config/auth_middleware.py
import jwt
from fastapi import HTTPException, Request, Depends
import os
from jwt import PyJWKClient

COGNITO_USER_POOL_ID = "us-east-1_JudMXeuR1"
COGNITO_REGION = "us-east-1"
COGNITO_ISSUER = f"https://cognito-idp.{COGNITO_REGION}.amazonaws.com/{COGNITO_USER_POOL_ID}"
COGNITO_JWKS_URL = f"{COGNITO_ISSUER}/.well-known/jwks.json"
COGNITO_CLIENT_ID = "92oh9nmejo0d5a5ga51il8u3o"  # Asegúrate de usar el correcto

ALGORITHM = "RS256"
jwks_client = PyJWKClient(COGNITO_JWKS_URL)

def get_public_key(header):
    signing_key = jwks_client.get_signing_key(header["kid"])
    return signing_key.key

async def verify_token(request: Request):
    auth_header = request.headers.get("Authorization")
    
    if not auth_header or not auth_header.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="No token provided")

    token = auth_header.replace("Bearer ", "")

    try:
        header = jwt.get_unverified_header(token)
        public_key = get_public_key(header)

        decoded_token = jwt.decode(
            token, 
            public_key, 
            algorithms=[ALGORITHM], 
            issuer=COGNITO_ISSUER,
            options={"verify_aud": False}  # 🔥 Opción agregada para evitar error de `aud`
        )

        print("✅ Token decodificado correctamente:", decoded_token)
        return decoded_token

    except jwt.ExpiredSignatureError:
        raise HTTPException(status_code=401, detail="Token expired")
    except jwt.InvalidTokenError as e:
        raise HTTPException(status_code=401, detail=f"Invalid token: {str(e)}")

File: BackEnd/Domains/Categories/DeleteCategory/src/config/authMiddleware.js
const jwt = require("jsonwebtoken");
const jwksClient = require("jwks-rsa");

const client = jwksClient({
  jwksUri: `https://cognito-idp.us-east-1.amazonaws.com/us-east-1_JudMXeuR1/.well-known/jwks.json`,
});

async function getKey(header, callback) {
  try {
    const key = await client.getSigningKey(header.kid);
    const signingKey = key.publicKey || key.rsaPublicKey;
    callback(null, signingKey);
  } catch (error) {
    console.error("❌ Error obteniendo la clave de firma:", error);
    callback(error);
  }
}

// Middleware de autenticación
function authMiddleware(req, res, next) {
  const authHeader = req.headers["authorization"];

  if (!authHeader) {
    return res.status(401).json({ error: "No token provided" });
  }

  const token = authHeader.replace("Bearer ", "");

  jwt.verify(token, getKey, { algorithms: ["RS256"] }, (err, decoded) => {
    if (err) {
      console.error("❌ Error verificando token:", err.message);
      return res.status(401).json({ error: "Invalid or expired token" });
    }
    
    // Almacenar la información del usuario en la solicitud
    req.user = decoded;
    next();
  });
}

module.exports = authMiddleware;
